### PR TITLE
Fix ASAN issue in lib/cstring.

### PR DESCRIPTION
**Pull request by matthewtlam**
Signed-off-by: Matthew Lam [matthew.lam.qwerty@gmail.com](mailto:matthew.lam.qwerty@gmail.com)

### CODE REVIEW

**Comment by matthewtlam** [comment]
For context, running the ASAN tool on one of our tests revealed an `initialization_order_fiasco` runtime error.

```
SanitizerError
AddressSanitizer: initialization-order-fiasco [third_party/p4lang_p4c/lib/json.h:110]:23 in JsonValue;
Details
=================================================================
==4779==ERROR: AddressSanitizer: initialization-order-fiasco on address 0x7f2eaf3ee3c0 at pc 0x7f2eaefdbe2b bp 0x7ffdb7f50ba0 sp 0x7ffdb7f50b98
READ of size 8 at 0x7f2eaf3ee3c0 thread T0
    #0 0x7f2eaefdbe2a in JsonValue [third_party/p4lang_p4c/lib/json.h:110]:23
    #1 0x7f2eaefdbe2a in __cxx_global_var_init.2 [third_party/p4lang_p4c/lib/json.cpp:36]:34
    #2 0x7f2eaefdbe2a in _GLOBAL__sub_I_json.cpp third_party/p4lang_p4c/lib/json.cpp
    #3 0x7f2ecbbade29 in call_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1be29) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #4 0x7f2ecbbadc8a in _dl_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1bc8a) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #5 0x7f2ecbb99659 in _dl_start_user (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x7659) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)

0x7f2eaf3ee3c0 is located 32 bytes before global variable '(anonymous namespace)::cache()::g_cache' defined in 'third_party/p4lang_p4c/lib/cstring.cpp' (0x7f2eaf3ee3e0) of size 56
  registered at:
    #0 0x55fa54125a58 in __asan_register_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:369]:3
    #1 0x55fa54126b6e in __asan_register_elf_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:352]:3
    #2 0x7f2eaef73247 in asan.module_ctor alloc_trace.cpp
    #3 0x7f2ecbbade29 in call_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1be29) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #4 0x7f2ecbbadc8a in _dl_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1bc8a) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #5 0x7f2ecbb99659 in _dl_start_user (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x7659) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)

0x7f2eaf3ee3c0 is located 0 bytes inside of global variable 'cstring::empty' defined in 'third_party/p4lang_p4c/lib/cstring.cpp' (0x7f2eaf3ee3c0) of size 8
  registered at:
    #0 0x55fa54125a58 in __asan_register_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:369]:3
    #1 0x55fa54126b6e in __asan_register_elf_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:352]:3
    #2 0x7f2eaef73247 in asan.module_ctor alloc_trace.cpp
    #3 0x7f2ecbbade29 in call_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1be29) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #4 0x7f2ecbbadc8a in _dl_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1bc8a) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #5 0x7f2ecbb99659 in _dl_start_user (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x7659) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)

SUMMARY: AddressSanitizer: initialization-order-fiasco [third_party/p4lang_p4c/lib/json.h:110]:23 in JsonValue
```

**Comment by matthewtlam** [comment]
@jonathan-dilorenzo, @smolkaj for visibility

**Comment by asl** [comment]
I actually do have patch for this downstream

**Comment by qobilidop** [comment]
Do we understand why this wasn't caught by the existing [sanitizer CI workflow on GitHub](https://github.com/p4lang/p4c/blob/main/.github/workflows/ci-ubuntu-20-sanitizer-nightly.yml)?

**Comment by asl** [comment]
> Do we understand why this wasn't caught by the existing [sanitizer CI workflow on GitHub](https://github.com/p4lang/p4c/blob/main/.github/workflows/ci-ubuntu-20-sanitizer-nightly.yml)?

Yes. Though I was able to reproduce it with asan on downstream project. The order of static initializers do matter. And this depends on linker.

**Comment by asl** [comment]
I submitted downstream solution in https://github.com/p4lang/p4c/pull/4866 that should fix the underlying problem

@matthewtlam will you please double-check it fixes asan issues for you?

**Comment by matthewtlam** [comment]
@asl I will double-check that #4866 fixes the ASAN issues on my side

**Comment by asl** [comment]
> Do we understand why this wasn't caught by the existing [sanitizer CI workflow on GitHub](https://github.com/p4lang/p4c/blob/main/.github/workflows/ci-ubuntu-20-sanitizer-nightly.yml)?

To add more on this. We might want to try `ASAN_OPTIONS=check_initialization_order=true:strict_init_order=true`. Though it might be overly-aggressive :)

**Comment by matthewtlam** [comment]
@asl Unfortunately, the ASAN check failed with the changes in #4866 as shown below. I believe we use the most strict form of checking for ASAN.

```
SanitizerError
AddressSanitizer: initialization-order-fiasco [third_party/p4lang_p4c/lib/json.h:110]:23 in JsonValue (//third_party/pins_infra/p4testgen/bmv2:p4testgen_bmv2_integration_test); see [http://go/sanitizer-blaze-warning]
Details
=================================================================
==1324==ERROR: AddressSanitizer: initialization-order-fiasco on address 0x7fd667be8800 at pc 0x7fd6677baceb bp 0x7ffe123e3380 sp 0x7ffe123e3378
READ of size 8 at 0x7fd667be8800 thread T0
    #0 0x7fd6677bacea in JsonValue [third_party/p4lang_p4c/lib/json.h:110]:23
    #1 0x7fd6677bacea in __cxx_global_var_init.2 [third_party/p4lang_p4c/lib/json.cpp:36]:34
    #2 0x7fd6677bacea in _GLOBAL__sub_I_json.cpp third_party/p4lang_p4c/lib/json.cpp
    #3 0x7fd683b9ae29 in call_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1be29) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #4 0x7fd683b9ac8a in _dl_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1bc8a) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #5 0x7fd683b86659 in _dl_start_user (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x7659) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)

0x7fd667be8800 is located 32 bytes before global variable '(anonymous namespace)::cache()::g_cache' defined in 'third_party/p4lang_p4c/lib/cstring.cpp' (0x7fd667be8820) of size 8
  registered at:
    #0 0x558a85125c98 in __asan_register_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:392]:3
    #1 0x558a85126ece in __asan_register_elf_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:375]:3
    #2 0x7fd6677803b7 in asan.module_ctor alloc_trace.cpp
    #3 0x7fd683b9ae29 in call_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1be29) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #4 0x7fd683b9ac8a in _dl_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1bc8a) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #5 0x7fd683b86659 in _dl_start_user (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x7659) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)

0x7fd667be8800 is located 0 bytes inside of global variable 'cstring::empty' defined in 'third_party/p4lang_p4c/lib/cstring.cpp' (0x7fd667be8800) of size 8
  registered at:
    #0 0x558a85125c98 in __asan_register_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:392]:3
    #1 0x558a85126ece in __asan_register_elf_globals [third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_globals.cpp:375]:3
    #2 0x7fd6677803b7 in asan.module_ctor alloc_trace.cpp
    #3 0x7fd683b9ae29 in call_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1be29) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #4 0x7fd683b9ac8a in _dl_init (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x1bc8a) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)
    #5 0x7fd683b86659 in _dl_start_user (/usr/grte/v5/lib64/ld-linux-x86-64.so.2+0x7659) (BuildId: 3c4ff280c4f5e542ce622870f20d94f2)

SUMMARY: AddressSanitizer: initialization-order-fiasco [third_party/p4lang_p4c/lib/json.h:110]:23 in JsonValue
```

**Comment by asl** [comment]
> @asl Unfortunately, the ASAN check failed with the changes in #4866 as shown below. I believe we use the most strict form of checking for ASAN.

Ah, actually this is *different* issue!

#4866 fixes the problem when cstring cache is not live before any of cstring literals around.
#4865 fixes the problem where cstrings are not live before json

So, it looks like we need to use both. And actually reduce the number of static objects around, this makes p4c-as-a-library usage hard :(

File: lib/cstring.h [hunk 1]
```diff
@@ -18,11 +18,14 @@ limitations under the License.
 #define LIB_CSTRING_H_
 
 #include <cstddef>
+#include <cstdio>
```
Comment (asl) [inline lib/cstring.h:21]: why do we need this?

Comment (matthewtlam) [reply lib/cstring.h]: Removed marking as resolved

File: lib/cstring.h [hunk 2]
```diff
@@ -366,6 +368,9 @@ inline std::string &operator+=(std::string &a, cstring b) {
     return a;
 }
 
+inline cstring cstring::newline = cstring::literal("\n");
```
Comment (smolkaj) [inline lib/cstring.h:371]: Shall we combine declaration and definition and make them `static const inline`?

Comment (matthewtlam) [reply lib/cstring.h:371]: As discussed offline, it seems that cstring does not have a destructor and the compiler is not happy making cstring a const value as shown here 
`error: calling 'literal<const char, 2UL, void>' with incomplete return type 'cstring'
  274 |     static const inline cstring newline = literal("\n");`
Marking as resolved.

**Review by smolkaj — APPROVED** [review]
LGTM

**Pull request merged**